### PR TITLE
chore: support permissions boundaries in env stack IAM roles

### PR DESCRIPTION
**Pull request by huanjani**
Related: #3969, #2986.

This adds permissions boundaries to SSM and conditionally adds the permissions boundary field in EnvManager and Admin IAM roles in env stack templates.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/deploy/env.go [hunk 1]
```diff
@@ -45,7 +45,8 @@ type CreateEnvironmentInput struct {
 	RawMft             []byte                // Content of the environment manifest without any modifications.
 	ForceUpdate        bool
 
-	CFNServiceRoleARN string // Optional. A service role ARN that CloudFormation should use to make calls to resources in the stack.
+	CFNServiceRoleARN   string // Optional. A service role ARN that CloudFormation should use to make calls to resources in the stack.
+	PermissionsBoundary string // Optional name of permissions boundary for IAM Roles.
```
Comment (paragbhingre) [inline internal/pkg/deploy/env.go]:
```suggestion
	PermissionsBoundary string // Optional. A name of the permissions boundary for IAM Roles.
```

**Review by paragbhingre — COMMENTED** [review]
Just a nit, otherwise it looks good to me.

File: internal/pkg/deploy/app.go [hunk 2]
```diff
@@ -20,7 +20,7 @@ type CreateAppInput struct {
 	DNSDelegationAccounts []string          // Accounts to grant DNS access to for this application.
 	DomainName            string            // DNS Name used for this application.
 	DomainHostedZoneID    string            // Hosted Zone ID for the domain.
-	PermissionBoundary    string            // Name of the permission boundary for IAM roles.
+	PermissionsBoundary   string            // Name of the permissions boundary for IAM roles.
```
Comment (efekarakus) [inline internal/pkg/deploy/app.go]: nit:
```suggestion
	PermissionsBoundary   string            // Name of the IAM Managed Policy to set a permissions boundary.
```

File: internal/pkg/deploy/env.go (hunk 1 above)
Comment (efekarakus) [reply internal/pkg/deploy/env.go]:
```suggestion
	PermissionsBoundary string // Optional. An IAM Managed Policy name used as permissions boundary for IAM roles.
```

File: internal/pkg/template/templates/environment/partials/environment-manager-role.yml [hunk 3]
```diff
@@ -13,6 +13,9 @@ EnvironmentManagerRole:
         Principal:
           AWS: !Sub ${ToolsAccountPrincipalARN}
         Action: sts:AssumeRole
+{{- if .PermissionsBoundary }}
+    PermissionsBoundary: !Sub 'arn:${AWS::Partition}:iam::${AWS::AccountId}:policy/{{.PermissionsBoundary}}'
```
Comment (efekarakus) [inline internal/pkg/template/templates/environment/partials/environment-manager-role.yml:17]: Are we planning to update the other IAM roles in the env stack in a separate PR? For example: https://github.com/aws/copilot-cli/blob/4b45d6682865598b368819023f15185b6d572b93/internal/pkg/template/templates/environment/partials/custom-resources-role.yml#L2 

I think for that we will need to touch `env package` and `env deploy` as well.

Comment (huanjani) [reply internal/pkg/template/templates/environment/partials/environment-manager-role.yml:17]: Yup!

**Review by efekarakus — APPROVED** [review]

File: internal/pkg/deploy/cloudformation/stack/env.go [hunk 4]
```diff
@@ -63,7 +63,7 @@ var (
 	DefaultPrivateSubnetCIDRs   = []string{"10.0.2.0/24", "10.0.3.0/24"}
 )
 
-// NewEnvStackConfig return o CloudFormation stack configuration for deploying a brand new environment.
```
Comment (Lou1415926) [inline internal/pkg/deploy/cloudformation/stack/env.go:66]: :homer-disappearing

**Review by Lou1415926 — APPROVED** [review]
LGTM :shipit:

**Pull request merged**